### DISCUSSION TITLE
OpenAI integration with Elasticsearch

### DISCUSSION THREAD

**Discussion by Aleksa1996** [swirlai/swirl-search#1392] [Q&A]
Hello i have integrated Elasticsearch with swirl via search provider and its working flawlessly. Now i would like to use RAG response together with Elasticsearch search provider. Is it possible ? How can i do it ? I followed [AI Guide](https://docs.swirl.today/AI-Guide.html) but it's not working. I have set OPENAI_API_KEY as environment variable through docker compose. I also tried to add that variable manually inside container .env, but still does not work. Thanks in advance!

Swirl image: `swirlai/swirl-search:latest`

Logs i've managed to capture:
```
logs/celery-worker.log -> [2024-07-16 04:33:13,691: ERROR/ForkPoolWorker-6] federate[]: GenAIGPT_10_9: name 'llm' is not defined
```

**Comment by erikspears** [comment]
Hi @Aleksa1996 ,

Happy to try and help get you going with this.  If I read the above correctly, you have an Elastic SearchProvider configured, enabled, and working, which is great!  In order to include those results in RAG processing, there are a couple of next steps.

1. Do go ahead and setup your OpenAI key and the necessary MSAL values to support RAG in Docker (which you mentioned using) via the `export` commands outlined in the [Quick Start Guide](https://docs.swirl.today/Quick-Start.html#start-swirl-in-docker).
2. You then need to add configuration to the `page_fetch_config_json` section of your Elastic SearchProvider, as outlined in the [AI Guide](https://docs.swirl.today/AI-Guide.html#setting-up-rag).  Try the basic configuration first, like so:
```
"page_fetch_config_json": {
        "cache": "false",
        "headers": {
            "User-Agent": "Swirlbot/1.0 (+http://swirl.today)"
        },
        "timeout": 10
}, 
```

Be sure to PUT that change to the SearchProvider and double-check that it saved.  With those steps accomplished, refresh your Galaxy UI, and try a fresh search that gets results from your Elastic source.  Then, toggle on the "Generate AI Response" option just under the search box.  If you don't get a RAG response in a minute or less, grab any errors that have gone by in your Docker app window, and post them here for us to take a look at.

Many thanks,
--Erik
